tmitchell/exercise.py

```diff
@@ -0,0 +1,281 @@
+#!/usr/bin/python3
+"""Run an interactive shell, users can view and edit a phone book and then save it to file."""
+class Person:
+    """Entry in phonebook able tyo store and display information"""
+    def __init__(self, name, hnumber, mnumber):
+        self.name = name
+        self.hnumber = hnumber
+        self.mnumber = mnumber
+
+    def display(self):
+        """Print stored information to screen."""
+        print("Name: %s" % (self.name))
+        print("Home Number: %s" % (self.hnumber))
+        print("Mobile Number: %s" % (self.mnumber))
+
+    def edit(self, new_name):
+        """Request user input and use it to edit entry details"""
+        if new_name != "":
+            self.name = new_name
+        else:
+            print("No changes will be made to the entry name")
+        print("Current entry home number:")
+        print(self.hnumber)
+        print("Enter new home number: (leave blank to remain same)")
+        new_hnumber = input("")
+        if new_hnumber != "":
+            self.hnumber = new_hnumber
+        else:
+            print("No changes will be made to home phone number")
+        print("Current entry mobile number:")
+        print(self.mnumber)
+        print("Enter new mobile number: (leave blank to remain same)")
+        new_mnumber = input("")
+        if new_mnumber != "":
+            self.mnumber = new_mnumber
+        else:
+            print("No changes will be made to mobile phone number")
+        print("Revised entry:")
+        self.display()
+
+class Book:
+    """Stores list of Person objects, contains methods to manipulate and view that list."""
+    def __init__(self):
+        self.name = "phonebook"
+        self.people = []
+
+    def display_self(self):
+        """Iterate through all entries in book and display details."""
+        print("")
+        print("---Fonbuk 5000 User List---")
+        print("")
+        for i in self.people:
+            i.display()
+            print("")
+        print("---End of Fonbuk---")
+        print("")
+
+    def search(self):
+        """Request criteria from user, iterate through all entries and display matching results."""
+        print("Enter part or whole of entry name:")
+        search = input("")
+        count = 0
+        for i in self.people:
```

Comment:
Challenge: can you think of a way to make it run faster if you say, have 100k people?